src/http.cc

```diff
@@ -826,11 +839,124 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// A protocol-comparing predicate for STL search algorithms.
+/// Helps to match a protocol versioned name to a name without
+/// version (for example, 'websocket/1.2' should match to 'websocket' ).
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) : ref(s), refLen(len) {
+        const char *p = std::find(s, s + len, '/');
+        refBasicLen = p - s;
+        refIsVersioned = (refBasicLen !=  len);
+    }
+
+    bool operator() (const SBuf &checking) {
+        const bool checkingIsVersioned = (checking.find('/') == SBuf::npos);
+        if (checkingIsVersioned && !refIsVersioned)
+            return checking.caseCmp(ref, refBasicLen) == 0;
+        else
+            return checking.caseCmp(ref, refLen) == 0;
+    }
+
+private:
+    const char *ref; ///< The reference protocol
+
+    /// The full length of reference protocol string, including the
+    /// version part
+    size_t refLen;
+
+    /// The length of reference protocol string without version part
+    size_t refBasicLen;
+
+    /// \return true if the reference protocol includes version info
+    bool refIsVersioned;
+};
+
+bool
+HttpStateData::upgradeProtocolsSupported(const HttpReply *reply) const
+{
+    if (!upgradeProtocols)
+        return false;
+
+    const String upgradeProtos = reply->header.getList(Http::HdrType::UPGRADE);
+    const char *pos = nullptr;
+    const char *item;
+    int ilen = 0;
+    while (strListGetItem(&upgradeProtos, ',', &item, &ilen, &pos)) {
+        auto it = std::find_if(upgradeProtocols->cbegin(), upgradeProtocols->cend(), MatchProtocol(item, ilen));
+        if (it == upgradeProtocols->cend()) { // protocol not listed by client!
```

Comment:
Please initialize 'it' as const auto if possible.

Reply:
ok